Library/Homebrew/utils.rb

```diff
@@ -770,3 +770,39 @@ def number_readable(number)
   (numstr.size - 3).step(1, -3) { |i| numstr.insert(i, ",") }
   numstr
 end
+
+# True if this version of Ruby supports text encodings in its strings
+def ruby_has_encoding?
+  String.method_defined?(:force_encoding)
+end
+
+# Truncates a text string to fit within a byte size constraint,
+# preserving character encoding validity. The returned string will
+# be not much longer than the specified max_bytes, though the exact
+# shortfall or overrun may vary.
+def truncate_text_to_approximate_size(s, max_bytes, options = {})
+  front_weight = options.fetch(:front_weight, 0.5)
+  return s if s.bytesize <= max_bytes
+
+  if front_weight < 0.0 || front_weight > 1.0
+    raise "opts[:front_weight] must be between 0.0 and 1.0"
+  end
+  glue = "\n[...snip...]\n"
+  max_bytes_in = [max_bytes - glue.bytesize, 1].max
+  if ruby_has_encoding?
+    bytes = s.dup.force_encoding("BINARY")
+    glue_bytes = glue.encode("BINARY")
+  else
+    bytes = s
+    glue_bytes = glue
+  end
```

Comment:
If I can convince you to continue to use `s` and `glue` beyond this point, then the `else` could be omitted and this block could become:

```ruby
if ruby_has_encoding?
  s.force_encoding("BINARY")
  glue.encode!("BINARY")
end
```

(I guess one reason not to do this would be that this will modify `s`, so this could cause issues on the call site if this argument continues to be used there—and its encoding changes surprisingly.)